keps/sig-storage/20200319-selinux-relabeling.md

```diff
@@ -0,0 +1,477 @@
+    // may change label of all files on the volume to match the Pod requirements.
+    // This field must be "true" to enable "OnVolumeMount" value in Pod's SELinuxRelabelPolicy.
+    // Volumes provided by such CSI driver must be independent, in the sense that all volumes
+    // can be mounted to a single node with a different SELinux context in "-o context=<context>".
+    // mount option.
+    //
+    // When "false", Kubernetes won't pass any special SELinux mount options nor change
+    // SELinux labels on files on the volume, even when pod's
+    // podSecurityContext.seLinuxRelabelPolicy is "OnVolumeMount".
+    //
+    // When not set (default), Kubernetes won't pass any special SELinux mount options
+    // when "podSecurityContext.seLinuxRelabelPolicy ="OnVolumeMount".
+    // It will autodetect if a volume supports SELinux or not by checking for "seclabel"
+    // mount option on a NodePublished directory. If "seclabel" is present (i.e. kernel supports
+    // SELinux labels for this volume", container runtime may change label of all files on the
+    // volume to match the Pod requirements.
+    SupportsSELinux *bool;
+    ...
+}
+
+// For context:
+type CSIDriver struct {
+    Spec CSIDriverSpec
+}
+```
+
+### Implementation Details/Notes/Constraints [optional]
+
+#### `mount -o context`
+Linux kernel, with SELinux compiled in, allows `mount -o context=s0:c309,c383 <what> <where>` to mount a volume and pretend that all files on the volume have given SELinux label.
+It works only for the first mount of the volume!
+It does not work for bind-mounts or any subsequent mount of the same volume.
+
+Note that volumes mounted with `-o context` don't have `seclabel` in their mount options.
+In addition, calling `chcon` there will fail with `Operation not supported`.
+
+### New Kubernetes behavior
+
+* If kubelet *knows* SELinux context of a pod / container to run (i.e. Pod/Container contains at least `SELinuxOptions.Level`):
+    * And pod's `SELinuxRelabelPolicy` is `OnVolumeMount`:
+        * And if the in-tree volume plugin supports SELinux / `CSIDriver.Spec.SupportsSELinux` is explicitly `true`:
+            * Kubelet tries to mount the volume for the Pod with given SELinux label using `mount -o context=XYZ`.
+                * Kubelet makes sure the option is passed to the first mount in all in-tree volume plugins (incl. ephemeral volumes like Secrets).
+                * Kubelet passes it as a mount option to all CSI calls for given volume.
+            * After the volume is mounted, kubelet checks that the root of the volume has the expected SELinux label, i.e. that the volume was mounted correctly.
+                * If the volume root has expected label, kubelet passes the volume to the container runtime without any ":z" or ":Z" options - no relabeling is necessary.
+                * If the volume root has unexpected label, for example when CSI driver did not apply `-o context` correctly, or the volume was already mounted:
```

Comment:
Do we want to support this fallback behavior? Could this hide bugs in the driver?

What is the scenario where the volume is already mounted but has the wrong label?

Reply:
This fallback comes from KEP version without a field in CSIDriver, so there had to be fallback for faulty CSI drivers.

Thinking again about this and @gnufied's comment above. If a CSI driver explicitly says it supports SELinux, it's a driver bug if the volume root has unexpected label. Pod setup should fail. I rephrased the bullets.